libcloud/compute/drivers/vsphere.py

```diff
@@ -49,506 +34,206 @@
-        if ex_remove_files:
-            request = VI.Destroy_TaskRequestMsg()
-
-            _this = request.new__this(vm._mor)
-            _this.set_attribute_type(vm._mor.get_attribute_type())
-            request.set_element__this(_this)
-            ret = server._proxy.Destroy_Task(request)._returnval
-            task = VITask(ret, server)
-
-            # Wait for the task to finish
-            status = task.wait_for_state([task.STATE_SUCCESS,
-                                          task.STATE_ERROR])
-
-            if status == task.STATE_ERROR:
-                raise LibcloudError('Error destroying node: %s' %
-                                    (task.get_error_message()))
+    def __init__(self, host, username, password, verify_ssl=True):
+        if verify_ssl:
+            context = ssl.create_default_context()
         else:
-            request = VI.UnregisterVMRequestMsg()
-
-            _this = request.new__this(vm._mor)
-            _this.set_attribute_type(vm._mor.get_attribute_type())
-            request.set_element__this(_this)
-            ret = server._proxy.UnregisterVM(request)
-            task = VITask(ret, server)
-
-        return True
+            context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+            context.verify_mode = ssl.CERT_NONE
 
-    @wrap_non_libcloud_exceptions
-    def ex_stop_node(self, node):
-        vm = self._get_vm_for_node(node=node)
-        vm.power_off()
+        try:
+            self.conn = connect.SmartConnect(host=host,
+                                             user=username,
+                                             pwd=password,
+                                             sslContext=context)
+            atexit.register(connect.Disconnect, self.conn)
+        except Exception as e:
+            err = str(e).lower()
+            if "incorrect user name" in err:
+                raise InvalidCredsError("username/password not acepted")
+            if "connection refused" in err or "is not a vim server" in err:
+                raise Exception("host doesn't appear to be a vSphere instance")
+            if "name or service not known" in err:
+                raise Exception("couldn't connect to '{}'".format(host))
+            if "certificate verify failed" in err:
+                raise Exception("SSL certificate couldn't be verified; if "
+                                "this is expected `verify_ssl=False` can "
+                                "be added to the driver to suppress checks")
+            raise Exception("unexpected error: {}".format(e))
+
+    @property
+    def _root(self):
+        return self.conn.RetrieveContent().rootFolder
+
+    @property
+    def _datacenters(self):
+        return [dc for dc in self._root.childEntity
+                if type(dc) is vim.Datacenter]
+
+    @property
+    def _vms(self):
+        vms = []
+        nodes = [n for dc in self._datacenters
+                 for n in dc.vmFolder.childEntity]
```

Comment:
Looks like you only go through this list once, so you could make it a generator:
```python
        nodes = (n for dc in self._datacenters
                 for n in dc.vmFolder.childEntity)